Xmc4xxx interrupt controller

talih0 opened the pull request:
This main aim of this PR is to review interrupt controller driver.
In Infineon XMC4XXX SoCs, gpio interrupts are triggered via an
Event Request Unit (ERU) module. A subset of the gpios are
connected to the ERU. The ERU monitors edge triggers and creates
a SR.
    
The interrupt controller driver configures the ERU for a target port/pin combination
for rising/falling edge events. Note that the ERU module does
not generate SR based on the gpio level. Internally the ERU
tracks the *status* of an event. The status is set on a positive
edge and unset on a negative edge (or vice-versa depending on
the configuration). The value of the status is used to implement
a level triggered interrupt; The ISR checks the status flag and
calls the callback function if the status is set.

The ERU configurations for supported port/pin combinations are
stored in a devicetree file dts/arm/infineon/xmc4xxx_x_x-intc.dtsi.
The configurations are stored in the opaque array
uint16 port_line_mapping[].

talih0 committed:
drivers: interrupt_controller: Add XMC4XXX ERU driver

In Infineon XMC4XXX SoCs, gpio interrupts are triggered via an
Event Request Unit (ERU) module. A subset of the gpios are
connected to the ERU. The ERU monitors edge triggers and creates
a SR.

This driver configures the ERU for a target port/pin combination
for rising/falling edge events. Note that the ERU module does
not generate SR based on the gpio level. Internally the ERU
tracks the *status* of an event. The status is set on a positive
edge and unset on a negative edge (or vice-versa depending on
the configuration). The value of the status is used to implement
a level triggered interrupt; The ISR checks the status flag and
calls the callback function if the status is set.

The ERU configurations for supported port/pin combinations are
stored in a devicetree file dts/arm/infineon/xmc4xxx_x_x-intc.dtsi.
The configurations are stored in the opaque array
uint16 port_line_mapping[].

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
drivers: gpio_xmc4xxx: Use interrupt controller for edge/level interrupts

Use the XMC4XXX interrupt controller for edge/level gpio interrupts.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

galak commented on drivers/gpio/Kconfig.xmc4xxx:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2022 Schlumberger
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_INFINEON_XMC4XXX_GPIO := infineon,xmc4xxx-gpio
```
Look at new Kconfig style in https://github.com/zephyrproject-rtos/zephyr/pull/48085

galak commented on dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml:
```diff
@@ -0,0 +1,21 @@
+    "#gpio-cells":
+      const: 2
+
+    port-id:
```
Can't you easily "compute" `port-id` by doing simple math on the the REG_ADDR?  as such, I think you should remove `port-id` and just make a macro in the code to do the math.

talih0 replied:
> Can't you easily "compute" `port-id` by doing simple math on the the REG_ADDR? as such, I think you should remove `port-id` and just make a macro in the code to do the math.

yes true, will update.

talih0 replied:
I removed port-id binding in the latest version and used DT_HAS_xx_ENABLED for gpio and interrupt controller.

galak reviewed (dismissed):
Some Kconfig and dts updates requested.

Made the comment on gpio/Kconfig.xmc4xxx, but also applies to interrupt-controller/Kconfig.xmc4xxx.

cfriedt commented:
Can I suggest keeping this PR in draft state, since it's dependent on multiple PRs being merged first?

cfriedt commented:
This PR probably shouldn't be assigned to the GPIO maintainer. However, the interrupt-controller driver area is lacking a maintainer currently. Any suggestions?

talih0 commented:
ping

MaureenHelm commented:
Needs rebase

talih0 commented:
> Needs rebase

Thanks, I've rebased. I also added a fix for a special case where the gpio level was set active before the interrupt was configured.

galak commented on dts/arm/infineon/xmc4500_F100x1024-intc.dtsi:
```diff
@@ -0,0 +1,45 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* File auto-generated with python script */
```
what is auto generating this file?

talih0 replied:
It's my own python script. I'll remove the comment.

galak commented:
> This PR probably shouldn't be assigned to the GPIO maintainer. However, the interrupt-controller driver area is lacking a maintainer currently. @galak - any suggestions?

No suggestion.

talih0 commented:
> > This PR probably shouldn't be assigned to the GPIO maintainer. However, the interrupt-controller driver area is lacking a maintainer currently. @galak - any suggestions?
> 
> No suggestion.

cfriedt reviewed (dismissed):
GPIO changes look fine

cfriedt commented on drivers/gpio/gpio_xmc4xxx.c:
```diff
@@ -173,6 +211,9 @@ static const struct gpio_driver_api gpio_xmc4xxx_driver_api = {
 	.port_clear_bits_raw = gpio_xmc4xxx_clear_bits_raw,
 	.port_toggle_bits = gpio_xmc4xxx_toggle_bits,
 	.pin_interrupt_configure = gpio_xmc4xxx_pin_interrupt_configure,
+#if defined(CONFIG_XMC4XXX_INTC)
```
Here, you could use `IS_ENABLED(CONFIG_XMC4XXX_INTC) ? manage_callback : NULL` which would require fewer `#ifdef`s

talih0 replied:
thanks, I've updated the PR.

cfriedt reviewed (dismissed):
Just a general comment to optionally reduce `#ifdef`s. Otherwise the GPIO code looks fine to me

carlescufi reviewed (approved):
LGTM

talih0 commented:
@carlescufi, @cfriedt, thank you for reviewing
